### PR TITLE
Improve slow term ordering queries

### PR DESCRIPTION
**Pull request by mikejolley**
Whilst testing performance I found some slow sorting methods which we can avoid with some newer filters.

Background; attributes and product cats can be sorted into a custom order. This custom sort param is stored in term meta. We previously had a custom param (menu_order) you could pass into get_terms to sort terms. We also tried to apply this default if nothing was passed. This lead to all term queries for some taxonomies to join termmeta in order to sort by the order param.

Lots of this was legacy also - back when wp_termmeta did not exist. We don't need that legacy code any more.

This PR does the following:

- Removes legacy woocomerce_termmeta table which is unused, and swaps out all calls to legacy term meta functions to the new WP ones.
- Rather than try to apply term sorting on the fly if no params are sent, this applies our sorting setting as the default sort order via the `get_terms_defaults` hook
- Use `pre_get_terms` hook to detect `menu_order` sorting param and apply to the term query. Meta sorting is now possible through WP.
- Use `terms_clauses` hook to support numeric name sorting and to improve the menu order sorting to work if meta is missing.

Because sorting is applied by default using the hooks above, we no longer need to sort the results of get_terms using manual php based sorting functions so I removed those also.

Things to test:
- Product categories. Drag and drop sort in the backend and ensure the results show on the frontend on shop page or in product cat widgets.
- Product attributes. Same as above, but also check attributes are sorted on variable product pages.
- On the backend cats/attribute screens, click a table heading to sort by something else like name. This should override our default and sort correctly.

Note; there is an update routine to migrate old meta to a consistent `order` meta key. We had things like `order_pa_color` before because terms could share meta data, but WP implemented term splitting so this was no longer needed.

### CODE REVIEW

**Comment by rodrigoprimo** [comment]
I skimmed through this PR and here are some initial thoughts:

- A few unit tests and two e2e tests are failing. I didn't check them but judging by the name of the tests that are failing, it seems related to the changes in this PR. Could you please take a look?
- For future PRs, could you please do PHPCS changes in their own commits? Example of a commit that includes coding standards fixes: https://github.com/woocommerce/woocommerce/pull/22570/commits/8675f3195d3e36262ca160ecb0a60032fee951d6. It is much faster to review PRs when the PHPCS changes are in their own commits. 
- I added a [commit](https://github.com/woocommerce/woocommerce/pull/22570/commits/c798ea9ea89b1d0b4ee3be454fce6263e29cf41e) to fix PHPCS errors on `includes/admin/class-wc-admin-taxonomies.php`.

Once the tests are passing, I can take a closer look and review this PR.

**Review by claudiulodro — DISMISSED** [review]
Seems like a reasonable change. Term meta has been in WP for years at this point, so I don't think anyone still needs the legacy tables. Code looked good, sorting worked fine in my testing, and unit tests pass.

Just curious, what sort of performance improvements did you record for this change (if any)?

Tagging this with 'Needs docs' so we remember to document the deprecations.

**Comment by mikejolley** [comment]
@claudiulodro I didn't keep a record :D It was shown in the profiler - the sorting was all PHP based before so was sorting on potentially thousands of results then returning a subset.

File: includes/wc-term-functions.php [hunk 1]
```diff
@@ -10,6 +10,101 @@
...
+		$terms_query->meta_query->parse_query_vars( $args );
+	}
+
+}
+add_action( 'pre_get_terms', 'wc_change_pre_get_terms', 10, 1 );
+
+/**
+ * Adjust term query to handle custom sorting parameters.
+ *
+ * @deprecated 3.6.0 No longer used as WP supports meta based sorting.
```
[the remark this answers is not in the document]
Comment (mikejolley) [reply includes/wc-term-functions.php]: Whoops, no I repurposed it in the end so it needs to stay hooked.

File: includes/wc-deprecated-functions.php [hunk 2]
```diff
@@ -1012,3 +1012,78 @@ function wc_get_core_supported_themes() {
 	wc_deprecated_function( 'wc_get_core_supported_themes()', '3.3' );
 	return array( 'twentyseventeen', 'twentysixteen', 'twentyfifteen', 'twentyfourteen', 'twentythirteen', 'twentyeleven', 'twentytwelve', 'twentyten' );
 }
+
+/**
+ * When a term is split, ensure meta data maintained.
+ *
+ * @param  int    $old_term_id      Old term ID.
```
Comment (claudiosanches) [inline includes/wc-deprecated-functions.php]: Missing `@deprecated` for those functions.

File: includes/admin/class-wc-admin-taxonomies.php [hunk 3]
```diff
@@ -107,13 +106,7 @@ public function create_term( $term_id, $tt_id = '', $taxonomy = '' ) {
 	 * @param mixed $term_id Term ID.
 	 */
 	public function delete_term( $term_id ) {
```
Comment (kloon) [inline includes/admin/class-wc-admin-taxonomies.php]: @deprecated missing?

**Review by kloon — CHANGES_REQUESTED** [review]
Looks good and works, just missing a `@deprecated` param on delete_meta function docblock

**Comment by mikejolley** [comment]
@kloon Docblock sorted and synced with master. Merge after re-review :)?

**Review by kloon — APPROVED** [review]

**Pull request merged**